lib/ads/media_tailor_ad_manager.js

```diff
@@ -432,34 +434,62 @@ shaka.ads.MediaTailorAdManager = class {
     let previousAd = false;
     if (this.ad_) {
       previousAd = true;
-      const firstQuartileTime = this.mediaTailorAd_.startTimeInSeconds +
-          0.25 * this.mediaTailorAd_.durationInSeconds;
-      const midpointTime = this.mediaTailorAd_.startTimeInSeconds +
-          0.5 * this.mediaTailorAd_.durationInSeconds;
-      const thirdQuartileTime = this.mediaTailorAd_.startTimeInSeconds +
-          0.75 * this.mediaTailorAd_.durationInSeconds;
-      if (!this.eventsSent.includes('firstQuartile') &&
-          currentTime >= firstQuartileTime) {
-        this.eventsSent.push('firstQuartile');
-        this.sendTrackingEvent_('firstQuartile');
-      } else if (!this.eventsSent.includes('midpoint') &&
-          currentTime >= midpointTime) {
-        this.eventsSent.push('midpoint');
-        this.sendTrackingEvent_('midpoint');
-      } else if (!this.eventsSent.includes('thirdQuartile') &&
-          currentTime >= thirdQuartileTime) {
-        this.eventsSent.push('thirdQuartile');
-        this.sendTrackingEvent_('thirdQuartile');
-      }
+      goog.asserts.assert(this.mediaTailorAd_, 'Ad should be defined');
+      this.sendInProgressEvents_(currentTime, this.mediaTailorAd_);
       const remainingTime = this.ad_.getRemainingTime();
       const duration = this.ad_.getDuration();
       if (duration > 0 && (remainingTime <= 0 || remainingTime > duration)) {
         this.onEnded_();
       }
-      if (this.ad_&& this.ad_.isLinear()) {
-        return;
-      }
     }
+    if (this.ad_&& this.ad_.isLinear()) {
```

Comment:
Hm. Looking at this function in the simplified state like this, it appears that this clause is no longer necessary? It's just the inverse of the next clause. You could rewrite this half of the function as...

```js
if (!this.ad_ || !this.ad_.isLinear()) {
  this.checkLinearAds_(currentTime);
  if (!this.ad_) {
    this.checkNonLinearAds_(currentTime);
  }
  if (previousAd && !this.ad_) {
    this.onAdBreakEnded_();
  }
}
```

Admittedly I'm not sure if that's more or less clear than the return-early.

Reply:
Changed!